File: ebuild/private/common/makechroot/binpkgs.go
// Copyright 2022 The ChromiumOS Authors.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package makechroot

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"cros.local/bazel/ebuild/private/common/fileutil"
	"cros.local/bazel/ebuild/private/common/portage/binarypackage"
)

func CopyBinaryPackages(packagesDir string, packagePaths []string) (atoms []string, err error) {
	const binaryExt = ".tbz2"

	for _, packagePath := range packagePaths {
		xp, err := binarypackage.ReadXpak(packagePath)
		if err != nil {
			return nil, fmt.Errorf("reading %s: %w", filepath.Base(packagePath), err)
		}
		category := strings.TrimSpace(string(xp["CATEGORY"]))
		pf := strings.TrimSpace(string(xp["PF"]))

		categoryDir := filepath.Join(packagesDir, category)
		if err := os.MkdirAll(categoryDir, 0o755); err != nil {
			return nil, err
		}

		copyPath := filepath.Join(categoryDir, pf+binaryExt)
		if err := fileutil.Copy(packagePath, copyPath); err != nil {
			return nil, err
		}

		atoms = append(atoms, fmt.Sprintf("=%s/%s", category, pf))
	}

	return atoms, nil
}

File: ebuild/private/common/makechroot/src/binpkgs.rs
// Copyright 2023 The ChromiumOS Authors.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use anyhow::Result;
use std::path::Path;

pub fn copy_binary_packages<P: AsRef<Path>>(
    packages_dir: &Path,
    package_paths: &[P],
) -> Result<Vec<String>> {
    const BINARY_EXT: &str = ".tbz2";

    let mut atoms = Vec::new();

    for package_path in package_paths {
        let xp = binarypackage::BinaryPackage::new(package_path)?.xpak()?;

        let category = std::str::from_utf8(&xp["CATEGORY"])?.trim();
        let pf = std::str::from_utf8(&xp["PF"])?.trim();

        let category_dir = packages_dir.join(category);
        std::fs::create_dir_all(&category_dir)?;

        let copy_path = category_dir.join(pf.to_owned() + BINARY_EXT);
        std::fs::copy(package_path, copy_path)?;

        atoms.push(format!("={}/{}", category, pf));
    }
    Ok(atoms)
}
